ext/ddtrace_profiling_native_extension/extconf.rb

```diff
@@ -192,6 +192,14 @@ def add_compiler_flag(flag)
   $defs << '-DUSE_LEGACY_RB_VM_FRAME_METHOD_ENTRY'
 end
 
+# On older Rubies, rb_gc_force_recycle allowed to free objects in a way that
+# would be invisible to free tracepoints, finalizers and without cleaning
+# obj_to_id_tbl mappings.
+$defs << '-DHAVE_WORKING_RB_GC_FORCE_RECYCLE' if RUBY_VERSION < '3.1'
+
+# On older Rubies, there was no RUBY_SEEN_OBJ_ID flag
+$defs << '-DNO_SEEN_OBJ_ID_FLAG' if RUBY_VERSION < '2.7'
```

Comment:
Minor: It's not a pattern we've used before on this file, but since these two defines will usually be used together

```c
#if defined(HAVE_WORKING_RB_GC_FORCE_RECYCLE) && ! defined(NO_SEEN_OBJ_ID_FLAG)
```

...I wonder if we should go for a `'-DNEED_GC_FORCE_RECYCLE_BUG_WORKAROUND' if RUBY_VERSION < '3.1' && RUBY_VERSION > '2.6'` or something like that?

Reply:
I feel this makes things more awkward in multiple senses:
* Feel like we reduce the maintenance burden of `extconf.rb` by trying to keep things more directly linked to runtime properties rather than our own "concepts" which people would need to dig deeper to determine conditions for (e.g. it's "straightforward" to determine whether a ruby version has a working `rb_gc_force_recycle` or has a `FL_SEEN_OBJ_ID` constant but less so to know whether it needs a `gc_force_recycle_bug_workaround` or what exactly that is).
* Some of the tests for this workaround rely on native functions registered in the `StackRecoder::Testing` module to proxy to runtime functionality (`rb_gc_force_recycle` and checking of `FL_SEEN_OBJ_ID`). By keeping the separate macros we can keep these functions registered with all Ruby versions and ensuring their results are consistent with Rubies that may be missing that functionality (by either doing no-ops or returning false in the case of flag checks). With a single workaround macro it feels like we'd need to either conditionally register these macros only in those cases where we know we'll be using them (which IMO adds a synchronization burden between a spec and its native code) or we'd have to tie the fallback functionality of those functions with less clear conditionals (e.g. why should `has_seen_id_flag` return false if `!NEED_GC_FORCE_RECYCLE_BUG_WORKAROUND`?).

So I'd vote for keeping things separate 😅 That being said, I did add a `CAN_APPLY_GC_FORCE_RECYCLE_BUG_WORKAROUND` macro to encode the conditions instead of having them in multiple places.

Reply:
+1 That's very reasonable :)